### DISCUSSION TITLE
[Question] SqsMessageProducer: Why is PublishAsync not awaited?

### DISCUSSION THREAD

**Discussion by tulior** [BrighterCommand/Brighter#3054] [Q&A] [answered]
Hi team,

I'm looking at the `SqsMessageProducer` class in the `paramore.brighter.messaginggateway.awssqs` project, and I have a question about the asynchronous behavior (or lack thereof) in the `Send` method.

The method uses `PublishAsync` from the AWS SDK to publish messages to SNS topics, which is great for potentially improving performance and responsiveness. However, it immediately calls `.Wait()` on the returned Task, effectively making the publishing synchronous.

I'm curious to understand the reasoning behind this design choice. Are there specific reasons why awaiting `PublishAsync` was not implemented?

Some potential benefits of using `await` and fully asynchronous publishing could include:

* **Improved scalability and resource utilization:** Avoiding blocking threads while waiting for publishing to complete could allow the application to handle more concurrent requests and use resources more efficiently.
* **Better responsiveness:** The application wouldn't be held up by potentially slow publishing operations, leading to a smoother user experience.

Of course, there might be valid reasons for the current implementation. Understanding the context behind the design choice would be helpful.

Thanks for your time and insights!

**Comment by iancooper** [comment, marked as answer]
I would suspect that the reason is that we have not historically bubbled async operations on producers up, as they are uncommon. For example although the Kafka transport supports an async/away compatible Produce, it is better to use the "sync" produce which buffers all produce requests. SQS is a little bit of an exception as it is an HTTP call.

As we are working on V10, we can review and see if it makes sense to expose async/await compatible operations on our producers.
